lib/ecto_ldap/adapter.ex

```diff
@@ -69,82 +91,7 @@ defmodule Ecto.Ldap.Adapter do
     end
   end
 
-  def execute(_repo, query_metadata, prepared, _params, preprocess, _options) do
-
-    search_response = search(prepared)
-    fields = ordered_fields(query_metadata.sources)
-    count = count_fields(query_metadata.select.fields, query_metadata.sources)
-
-    {:ok, {:eldap_search_result, results, []}} = search_response
-
-    result_set =
-      for entry <- results do
-        entry
-        |> process_entry
-        |> prune_attributes(fields)
-        |> generate_models(preprocess, count)
-      end
-    # IEx.pry
-
-    {count, result_set}
-  end
-
-  def ordered_fields(sources) do
-    {_, model} = elem(sources, 0)
-    model.__schema__(:fields)
-  end
-
-  def count_fields(fields, sources) do
-    Enum.map fields, fn
-      {:&, _, [idx]} = field ->
-        {_source, model} = elem(sources, idx)
-        {field, length(model.__schema__(:fields))}
-      field ->
-        {field, 0}
-    end
-  end
-
-  def process_entry({:eldap_entry, dn, attributes}) when is_list(attributes) do
-    List.flatten(
-      [dn: dn], 
-      Enum.map(attributes, fn {key, value} ->
-        {key |> to_string |> String.to_atom, value}
-      end))
-  end
-
-  def prune_attributes(attributes, fields) do
-    for field <- fields, do: Keyword.get(attributes, field)
-  end
-
-  def generate_models(row, preprocess, fields) do
-    Enum.map_reduce(fields, row, fn
-      {field, 0}, [h|t] ->
-        {preprocess.(field, h, nil), t}
-      {field, count}, acc ->
-        case split_and_not_nil(acc, count, true, []) do
-          {nil, rest} -> {nil, rest}
-          {val, rest} -> {preprocess.(field, val, nil), rest}
-        end
-    end) |> elem(0)
-  end
-
-  def split_and_not_nil(rest, 0, true, _acc), do: {nil, rest}
-  def split_and_not_nil(rest, 0, false, acc), do: {:lists.reverse(acc), rest}
-
-  def split_and_not_nil([nil|t], count, all_nil?, acc) do
-    split_and_not_nil(t, count - 1, all_nil?, [nil|acc])
-  end
-
-  def split_and_not_nil([h|t], count, _all_nil?, acc) do
-    split_and_not_nil(t, count - 1, false, [h|acc])
-  end
-
```

Comment:
Somehow I managed to forget most explanations in the commit message...

I moved around a bunch of methods so that the program-flow read easier and "private" methods could be found near their callers.